File: cifras/cifras.py
#!/usr/bin/python
"""
Simple math game: given a list of numbers, use the four basic 
operations (+, -, /, *)  to find (or be as close as possible to) 
another given number.

This code takes a functional (FP) approach. The Num class is used as
data container but no logic methods (more than repr) are implemented inside.

This problem was popularize first by the TV quiz-show "Des Chiffres et 
des Lettres" at France, and has travelled around ("Countdown" in the UK, 
"Cifras y Letras" in Spain, ...).

Author: Arnau Sanchez <tokland@gmail.com>
"""
import sys
import operator
import itertools

# Data model: an orthodox implementation would build a binary-tree,
# let's just save the string representation of the operation for simplicity.

class Num:
    """Number class with value and string representation."""
    def __init__(self, value, string=None):
        self.value = value
        self.string = string or str(value)
        
    def __repr__(self):
        return "Num(%s, %s)" % (str(self.value), repr(self.string))
            
# Functions
            
def makeop(operation, num1, num2):
    """Make operation between two Nums.
        
    >>> makeop(operator.mul, Num(2, "(1+1)"), Num(6, "(2*3)"))
    Num(12, '((1+1)*(2*3))')
    """
    op = operator
    operator_string = {op.add: "+", op.sub: "-", op.mul: "*", op.div: "/"}
    string = "(" + num1.string + operator_string[operation] + num2.string + ")"
    return Num(operation(num1.value, num2.value), string)     

def compact(it):
    return itertools.ifilter(bool, it)

def process_pair(num1, num2):
    """Yield all possible results from operating two nums. 
    
    Valid operations are: n1+2, n1*n2, n1-n2 (if n1 > n2), n2-n1 (if n2 > n1), 
        n1/n2 (if n1 is divisible by n2), n2/n1 (if n2 is divisible by n1).
    
    >>> list(process_pair(Num(7, "7"), Num(3, "3")))
    [Num(10, '(7+3)'), Num(4, '(7-3)'), Num(21, '(7*3)')]
    """
    num1a, num2a = ((num1, num2) if num1.value >= num2.value else (num2, num1))
    candidates = [
        (operator.add, None),
        (operator.sub, lambda x, y: x > y), 
        (operator.mul, None),
        (operator.div, lambda x, y: x % y == 0),
    ]
    # is it worth to remove duplicates?
    for (op, condition) in candidates:
        if not condition or condition(num1a.value, num2a.value):
            yield makeop(op, num1a, num2a)
                  
def process(ints):
    """Yield all possible numbers combinating pairs.
        
    >>> next(n for n in process([1,2,3,4,5,6]) if n.value == 576)
    Num(576, '(((4*(2+1))*6)*(5+3))')
    """
    def _process(nums):
        for i1, i2 in itertools.combinations(range(len(nums)), 2): 
            other_nums = nums[:i1] + nums[i1+1:i2] + nums[i2+1:]
            for num in process_pair(nums[i1], nums[i2]):
                yield num
                for x in _process([num] + other_nums):
                    yield x
    nums = map(Num, ints)
    for x in itertools.chain(nums, _process(nums)):
        yield x                     
            
def _test():
    """Run tests on docstrings"""
    import doctest
    return doctest.testmod(verbose=True)
        
def _main(args0):
    """Process options and arguments"""
    import optparse
    usage = "usage: cifras.py number1 ... numberN final\n\n" + __doc__.strip()
    parser = optparse.OptionParser(usage)
    parser.add_option('-t', '--test', dest='test', default=False,
        action="store_true", help='Run unittests')        
    parser.add_option('-v', '--verbose', dest='verbose', default=False,
        action="store_true", help='Enable verbose mode')        
    parser.add_option('-n', '--nsolutions', dest='nsolutions', default=1,
        type="int", help='Solutions to find (0 = all)')        
    options, args = parser.parse_args(args0)
    if options.test:
        return _test()
    elif len(args) < 2:
        parser.print_help()
        return 1    
    all_nums = map(int, args)
    nums, final = all_nums[:-1], all_nums[-1]
    solutions = (res for res in process(nums) if res.value == final)
    for index, result in enumerate(itertools.islice(solutions, options.nsolutions or None)):
         print "%s. %d = %s" % (index+1, result.value, result.string)
                
if __name__ == "__main__":
    sys.exit(_main(sys.argv[1:]))

File: cifras/cifras.hs
-- Based on a simple math game: given a list of numbers use the four basic 
-- operations (+, -, /, *)  between them to find (or be as close as possible to) 
-- another given number.

-- This is my first Haskell script.
-- Arnau Sanchez <pyarnau@gmail.com>

import Data.List
--import Text.Printf(printf)

type Value = Int

data BinaryOp = Add | Substract | Multiply | Divide
  deriving (Eq)
  
data ExtNum = Num { value :: Value } | 
              BinManip { value :: Value, num1 :: ExtNum, num2 :: ExtNum, op :: BinaryOp }
  deriving (Eq)

instance Show BinaryOp where
  show Add = "+"
  show Substract = "-"
  show Multiply = "*"
  show Divide = "/"
  
instance Show ExtNum where
  show (Num value) = show value
  show (BinManip value num1 num2 op) = "(" ++ show num1 ++ show op ++ show num2 ++ ")"

show2 x = (value x, show x)                                       
 
combinations :: Value -> [a] -> [[a]]
combinations 0 _  = [[]]
combinations n xs = [y:ys | y:xs' <- tails xs, ys <- combinations (n-1) xs']

combine2Nums :: ExtNum -> ExtNum -> [ExtNum]
combine2Nums n1 n2 = removeRepeated n1 n2 $ concat buildOps 
  where buildOps = [buildBin n1 n2 Add (+) True,
                    buildBin n1 n2 Substract (-) (value n1 > value n2),
                    buildBin n2 n1 Substract (-) (value n2 > value n1),
                    buildBin n1 n2 Multiply (*) True,
                    buildBin n1 n2 Divide div (value n1 `mod` value n2 == 0),
                    buildBin n2 n1 Divide div (value n2 `mod` value n1 == 0)]                  
        buildBin n1' n2' op f guard = 
          if guard then [BinManip ((value n1') `f` (value n2')) n1' n2' op] else []
        removeRepeated n1' n2' = filter (not . (`elem` [value n1', value n2']) . value)
                                      
getPairs :: (Eq a) => Value -> [a] -> [([a], [a])]
getPairs n xs = [(c, [x | x <- xs, not (x `elem` c)]) | c <- combinations n xs]

cifras :: [Value] -> [ExtNum]
cifras xs = concat (xs' : cifras2 [xs'])
  where xs' = [Num x | x <- xs]
    
cifras2 :: [[ExtNum]] -> [[ExtNum]]
cifras2 [] = [[]]
cifras2 xss = concat $ map fst pairs : (map cifras2 $ map joinPair pairs)
  where pairs = [(combine2Nums (xs !! 0) (xs !! 1), ys) | 
                 (xs, ys) <- concatMap (getPairs 2) xss]
        joinPair (xs, ys) = [x : ys | x <- xs]
                  
                 
main = do                  
  (putStrLn . show) $ length $ map show2 $ 
    filter ((== 765) . value) $ cifras [1, 3, 7, 10, 25, 50]
